HW1/P4/P4_BFS.py

```diff
@@ -0,0 +1,41 @@
+# Single-Source Breadth-First Search:
+# Takes 4 arguments:
+# - The graph in the form of an edgelist
+# - The starting superhero
+# - The maximum number of iteration in case there is an issue, should normally stop when no more nodes can be touched
+# - the spark context
+
+#It returns both the number of nodes touched INCLUDING the starting node
+# and an rdd (not collected) that saves the in the form of a list of tuples (Superhero touched, how far he is)
+
+def bfs(edgelist,SH,dmax,sc):
+    global accum 
+    accum = sc.accumulator(1)
+    
+    def increase_accum(x):
+        global accum
+        accum += 1
+        
+    dis = 1
+    
+    nextneigh = sc.parallelize([(SH,0)])
+    distance = sc.parallelize([(SH,0)])
+    
+    while dis <= dmax:
+        oldaccum = accum.value
+        
+        #Finding the next neighbours and excluding the ones that have been already visited:
+        nextneigh = edgelist.join(nextneigh).partitionBy(8).map(lambda (K,V): (V[0],V[1]+1)).distinct().subtractByKey(distance)
+        
```

Comment:
Nice implement. I used an algorithm that is similar to the traditional bfs. I guess it works because data for p4 is not very large. I should have figured out an algorithm like yours.